File: 백준/Silver/1926. 그림/그림.py
import sys
input = sys.stdin.readline

dy = [-1, 0, 1, 0]
dx = [0, 1, 0, -1]

def _input():
    N, M = map(int, input().split())
    board = []
    for _ in range(N):
        board.append(list(map(int, input().split())))
    return N, M, board

def dfs(y, x, board, visited):
    stack = []
    stack.append([y, x])

    cnt = 0
    while(stack):
        curY, curX = stack.pop()
        if(visited[curY][curX]): continue
        cnt += 1
        visited[curY][curX] = True

        for i in range(4):
            ny = curY + dy[i]
            nx = curX + dx[i]
            if(ny < 0 or len(board) <= ny or nx < 0 or len(board[0]) <= nx): continue
            if(board[ny][nx] == 0 or visited[ny][nx]): continue
            stack.append([ny, nx])
    
    return cnt


def solve():
    N, M, board = _input()
    visited = [[False for _ in range(M)] for _ in range(N)]

    answerCnt = 0
    answerSize = 0
    for i in range(N):
        for j in range(M):
            if(visited[i][j] or board[i][j] == 0): continue
            answerCnt += 1
            answerSize = max(answerSize, dfs(i,j,board,visited))

    print(answerCnt)
    print(answerSize)

solve()

File: 백준/Silver/1926. 그림/그림.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <cmath>

using std::ios;
using std::cout;
using std::cin;
using std::vector;
using std::queue;
using std::pair;

int N, M;
vector<vector<int>> board;
vector<vector<bool>> visited;

const int dy[4] = {-1, 0, 1, 0};
const int dx[4] = {0, 1, 0, -1};
int bfs(int y, int x) {
    int count = 0;
    queue<pair<int, int>> q;

    q.push({y, x});

    while(!q.empty()){
        pair<int, int> cur = q.front();
        q.pop();
        if(visited[cur.first][cur.second]) continue;
        visited[cur.first][cur.second] = true;
        ++count;

        for(int i = 0; i < 4; ++i){
            int ny = dy[i] + cur.first;
            int nx = dx[i] + cur.second;

            if(ny < 0 || N <= ny || nx < 0 || M <= nx) continue;
            if(visited[ny][nx]) continue;
            if(board[ny][nx] == 0) continue;

            q.push({ny, nx});
        }
    }
    return count;
}

void Input() {
    cin >> N >> M;
    board.assign(N, vector<int>(M));
    visited.assign(N, vector<bool>(M, false));
    for(auto& n : board){
        for(auto& m : n){
            cin >> m;
        }
    }
}

void Solve() {
    int answer = 0;
    int maxCnt = 0;
    for(int i = 0; i < N; ++i){
        for(int j = 0; j < M; ++j){
            if(visited[i][j]) continue;
            if(board[i][j] == 0) continue;
            ++answer;
            maxCnt = std::max(maxCnt, bfs(i, j));
        }
    }

    cout << answer << '\n';
    cout << maxCnt;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(nullptr);
    cout.tie(nullptr);
    
    Input();
    Solve(); 
    
    return 0;
}
